src/proper_target.erl

```diff
@@ -61,70 +61,159 @@
+-type target_state()  :: term().
+-type strategy_data() :: term().
+-type next_func()     :: fun ((target_state()) -> {target_state(), any()}).
+-type fitness_func()  :: fun ((target_state(), fitness()) -> target_state()) 
+                       | none.
+
+-type target()   :: {target_state(), next_func(), fitness_func()}.
+-type strategy() :: module().
+-type opts()     :: strategy() 
+                  | #{search_steps := integer(), search_strategy := strategy()}.
+
+-record(state,
+        {strategy           :: strategy(),
+         target = undefined :: target_state() | undefined,
+         data = undefined   :: strategy_data() | undefined}).
+-type state() :: #state{}.
+
+-export_type([fitness/0, tmap/0]).
+-export_type([target_state/0, strategy_data/0, next_func/0, fitness_func/0,
+              target/0, opts/0]).
 
--type target()    :: {target_state(), next_func(), fitness_func()}.
--type strategy()  :: module().
 
 %% -----------------------------------------------------------------------------
 %% proper_target callback functions for defining strategies
 %% ----------------------------------------------------------------------------
 
 %% strategy global initializer
--callback init_strategy(proper:setup_opts()) -> 'ok'.
-%% cleanup function
--callback cleanup() -> 'ok'.
+-callback init_strategy(proper:setup_opts()) -> strategy_data().
 %% target initializer
--callback init_target(tmap()) -> target().
-%% generator for shrinking
--callback get_shrinker(tmap()) -> proper_types:type().
-%% store, and retrieve state
--callback store_target(target_state()) -> 'ok'.
--callback retrieve_target() -> target() | 'undefined'.
+-callback init_target(tmap()) -> target_state().
+%% next function
+-callback next(target_state(), strategy_data()) -> 
+  {any(), target_state(), strategy_data()}.
 %% update the strategy with the fitness
--callback update_fitness(fitness()) -> 'ok'.
+-callback update_fitness(fitness(), target_state(), strategy_data()) -> 
+  {target_state(), strategy_data()}.
+%% reset strat
+-callback reset(target_state(), strategy_data()) -> 
+  {target_state(), strategy_data()}.
+
+
+%% -----------------------------------------------------------------------------
+%% API
+%% -----------------------------------------------------------------------------
+
+%% @doc Initializes targeted gen server based on a search strategy.
+%% Depending on the argument, the search steps will be taken from
+%% from the process dictionary or the setup options.
+
+-spec init_strategy(Opts :: opts()) -> {ok, pid()}.
+init_strategy(Strat) when is_atom(Strat) ->
+  Steps = get('$search_steps'),
+  init_strategy(#{search_steps => Steps, search_strategy => Strat});
+init_strategy(#{search_steps := Steps, search_strategy := Strat}) ->
+  Strategy = strategy(Strat),
+  proper_gen_next:init(),
+  Data = Strategy:init_strategy(#{numtests => Steps}),
+  gen_server:start_link({local, ?MODULE}, ?MODULE, [{Strategy, Data}], []).
```

Comment:
Is this really intended to be a singleton? Wouldn't two concurrent calls to `quickcheck` corrupt this gen_server's state?

Reply:
This could be fixed by using the servers `pid`, if I am not wrong. In this case however the `?MAXIMIZE` macro would be broker, as it has no way of knowing the `pid` of the current `proper_target gen_server`. 

I don't know if there is a solution to this.

Reply:
Yes, PropEr stores all run specific information there anyways and that makes it consistent and parallel runs work as well.

Reply:
Just saw that you already did this :-)

Reply:
@TheGeorge Great. I checked that the same pattern was followed with the typeserver, so I already made the necessary changes.